src/youtubeStream.py

```diff
@@ -0,0 +1,123 @@
+            LOGGER.error("Missing required attributes in Youtube stream Configuration.")
+            raise ValueError(f"Missing required attributes in Youtube stream Configuration: {', '.join(missing_attrs)}")
+        return
+    
+    def reconfigure(self, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]):
+        LOGGER.info("Reconfiguring " + self.name)
+        attrs = struct_to_dict(config.attributes)
+        self.video_url = str(attrs.get("video_url", self.default_url))
+
+        self.config = config
+        self.dependencies = dependencies
+
+        # Release the previous video capture object
+        if self.video_cap:
+            self.video_cap.release()
+
+        # Fetch the actual video URL using yt_dlp
+        stream_url = self.get_video_url(self.video_url)
+
+        # Init. the video capture object with the fetched stream URL
+        self.video_cap = cv2.VideoCapture(stream_url)
+    
+    def get_video_url(self, youtube_url: str) -> str:
+        # Fetch the direct video URL from YouTube using yt_dlp.
+        ydl_opts = {
+            'format': 'best',
+            'quiet': True,
+            'no_warnings': True,
+            'noplaylist': True,
+        }
+        ydl = YoutubeDL(ydl_opts)
+        try:
+            info = ydl.extract_info(youtube_url, download=False)
+            stream_url = info['url']
+            LOGGER.info(f"Fetched video stream URL: {stream_url}")
+            return stream_url
+        except Exception as e:
+            LOGGER.error(f"Error fetching video stream URL: {e}")
+            raise RuntimeError("Failed to fetch the YouTube stream URL.")
+        
+    async def get_image(self, mime_type: str = "", *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> ViamImage:
+        current_time = time.time()
+        
+        for i in range(5):  # Retry getting the video capture up to 10 times
+            ret, frame = self.video_cap.read()
+            if ret:
+                break
+            time.sleep(1)  # Wait 1 second before retrying 
+        if not ret:
+            raise RuntimeError("Failed to capture any frame from the video.")
+                
+        self.last_call_time = current_time
+
+        ret, jpeg = cv2.imencode('.jpg', frame)
+        if not ret:
+            raise RuntimeError("Failed to encode frame as JPEG.")
+        
+
+        image_data = jpeg.tobytes()
+        return ViamImage(image_data, CameraMimeType.JPEG)
+    
+    async def get_images(self, *, timeout: Optional[float] = None, **kwargs) -> Tuple[List[NamedImage], ResponseMetadata]:
+        raise NotImplementedError()
+
+    async def get_point_cloud(
+        self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs
+    ) -> Tuple[bytes, str]:
+        raise NotImplementedError()
+        
+    async def get_properties(self, *, timeout: Optional[float] = None, **kwargs) -> 'Camera.Properties':
```

Comment:
Is the return type of this method correct? And can you check if we can return any of the properties from this camera? I expect no for point clouds, intrinsics and distortion parameters because we can't point the stream at an object and run a calibration script. I forgot if there were other properties returned.

Reply:
okay from my research there is also a mime_type which can be returned. the cv2.VideoCap returns a jpeg and yt-dlp typically returns a mp4, so I included both of those in the get_properties return now.

Reply:
get_image endpoint always returns image/jpeg so changing it to just jpeg